src/AutoRest.CSharp/Common/Generation/Writers/SerializationWriter.cs

```diff
@@ -176,62 +189,278 @@ private static void WriteXmlDeserialize(CodeWriter writer, TypeDeclarationOption
                     {
-                        writer.Line($"return null;");
+                        writer.Line($"{typeof(Argument)}.{nameof(Argument.AssertNotNull)}(response, nameof(response));");
+                    }
+                    else
+                    {
+                        using (writer.Scope($"if (response is null)"))
+                        {
+                            writer.Line($"return null;");
+                        }
+                    }
+                    writer.Line();
+
+                    if (isXmlWireFormat)
+                    {
+                        writer.Line($"return {deserializeMethod}({typeof(XElement)}.{nameof(XElement.Load)}(response.{nameof(Response.ContentStream)}), {DefaultWireOptionsString});");
+                    }
+                    else
+                    {
+                        writer.Line($"using {typeof(JsonDocument)} doc = {typeof(JsonDocument)}.{nameof(JsonDocument.Parse)}(response.{nameof(Response.ContentStream)});");
+                        writer.Line($"return {deserializeMethod}(doc.{nameof(JsonDocument.RootElement)}, {DefaultWireOptionsString});");
                     }
                 }
+                writer.Line();
+            }
+        }
+
+        private static void WriteIModelJsonSerializableDeserialize(CodeWriter writer, CSharpType modelJsonSerializableType, string deserializeMethod, bool writeAsObject)
+        {
+            var modelType = writeAsObject ? typeof(object) : modelJsonSerializableType.Arguments[0];
+            var interfaceType = writeAsObject ? typeof(IModelJsonSerializable<object>) : modelJsonSerializableType;
 
-                var discriminator = serialization.Discriminator;
+            using (writer.Scope($"{modelType} {interfaceType}.{nameof(IModelJsonSerializable<object>.Deserialize)}(ref {typeof(Utf8JsonReader)} reader, {typeof(ModelSerializerOptions)} options)"))
+            {
+                WriteValidateFormat(writer, modelJsonSerializableType.Arguments[0]);
+                writer.Line($"using var doc = {typeof(JsonDocument)}.{nameof(JsonDocument.ParseValue)}(ref reader);");
+                writer.Line($"return {deserializeMethod}(doc.RootElement, options);");
+            }
+            writer.Line();
+        }
 
-                if (discriminator is not null && discriminator.HasDescendants)
+        private static void WriteInternalDeserialize(CodeWriter writer, TypeDeclarationOptions declaration, JsonObjectSerialization serialization, CSharpType modelSerializableType, bool isUnknownDerivedType, string deserializeMethod)
+        {
+            FormattableString signatureLine = $"internal static {modelSerializableType.Arguments[0]} {deserializeMethod}({typeof(JsonElement)} element, {typeof(ModelSerializerOptions)} options = default)";
+            if (isUnknownDerivedType)
+            {
+                writer.Line($"{signatureLine} => Deserialize{declaration.Name.Remove(0, 7)}(element, options);");
```

Comment:
The expression `declaration.Name.Remove(0, 7)` is fragile because it depends on the hard-coded word `Unknown`.
If we would like to remove the word unknown from its name, I assume we want to call the deserialization method on its base class, therefore could we do this:
```suggestion
                writer.Line($"{signatureLine} => Deserialize{serialization.ObjectType!.BaseType.Declaration.Name}(element, options);");
```
I believe it might be better if every `ObjectType` could have a property to store the name of its deserialization method. If we have that, we could have:
```suggestion
                writer.Line($"{signatureLine} => {serialization.ObjectType!.BaseType.DeserializationMethodName}(element, options);");
```